File: zmessaging/src/main/java/com/waz/api/Message.java
/*
 * Wire
 * Copyright (C) 2016 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.api;

import android.os.Parcel;
import android.os.Parcelable;
import com.waz.service.ZMessaging$;

public interface Message extends UiObservable, Parcelable {

    Parcelable.Creator<Message> CREATOR = new Parcelable.Creator<Message>() {
        @Override
        public Message createFromParcel(Parcel source) {
            return ZMessaging$.MODULE$.currentUi().messages().cachedOrFromParcel(source);
        }

        @Override
        public Message[] newArray(int size) {
            return new Message[size];
        }
    };

    enum Status {
        DEFAULT, SENT, PENDING, DELETED, FAILED, FAILED_READ, DELIVERED;

        public boolean isFailed() {
            return this == FAILED;
        }
    }

    enum Type {
        TEXT, TEXT_EMOJI_ONLY,
        //TODO rename to IMAGE_ASSET
        ASSET, ANY_ASSET, VIDEO_ASSET, AUDIO_ASSET,

        KNOCK, MEMBER_JOIN, RICH_MEDIA, LOCATION, RECALLED,

        MEMBER_LEAVE, CONNECT_REQUEST, CONNECT_ACCEPTED, RENAME,

        MISSED_CALL, SUCCESSFUL_CALL,

        OTR_ERROR, OTR_IDENTITY_CHANGED, OTR_VERIFIED, OTR_UNVERIFIED, OTR_DEVICE_ADDED, OTR_MEMBER_ADDED,

        STARTED_USING_DEVICE, HISTORY_LOST,

        UNKNOWN
    }

    interface Part {
        enum Type {
            TEXT, TEXT_EMOJI_ONLY, ASSET, ANY_ASSET, YOUTUBE, SOUNDCLOUD, TWITTER, SPOTIFY, WEB_LINK, GOOGLE_MAPS
        }
    }

    String getId();

    ImageAsset getImage();

    /**
     * Returns users (including self) that like this message.
     */
    User[] getLikes();


    boolean isEphemeral();

    boolean isExpired();

    /**
     * Retry sending a previously failed message.
     */
    void retry();
}

File: zmessaging/src/main/scala/com/waz/api/impl/Message.scala
/*
 * Wire
 * Copyright (C) 2016 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.api.impl

import com.waz.ZLog.ImplicitTag._
import com.waz.ZLog._
import com.waz.api
import com.waz.api.Message.Status
import com.waz.api.{EphemeralExpiration, UpdateListener}
import com.waz.model._
import com.waz.service.messages.MessageAndLikes
import com.waz.threading.Threading
import com.waz.ui._
import com.waz.utils.events.Signal

class Message(val id: MessageId, var data: MessageData)(implicit context: UiModule) extends api.Message with SignalLoading with UiObservable {

  private val convId = if (data == MessageData.Empty) Signal[ConvId]() else Signal(data.convId)
  private var lastMessageFromSelf = false
  private var lastMessageFromOther = false

  def this(msg: MessageAndLikes)(context: UiModule) = this(msg.message.id, msg.message)(context)
  def this(id: MessageId)(context: UiModule) = this(id, MessageData.Empty)(context)

  reload() // always reload because data from constructor might always be outdated already
  addLoader(zms => convId.flatMap(c => zms.messagesStorage.lastMessageFromSelfAndFromOther(c)), (Option.empty[MessageData], Option.empty[MessageData])) { case (fromSelf, fromOther) =>
    val isMostRecentFromSelf = fromSelf.exists(_.id == id)
    val isMostRecentFromOther = fromOther.exists(_.id == id)
    if (isMostRecentFromSelf != lastMessageFromSelf || isMostRecentFromOther != lastMessageFromOther) {
      lastMessageFromSelf = isMostRecentFromSelf
      lastMessageFromOther = isMostRecentFromOther
      notifyChanged()
    }
  }

  def reload() = context.zms.flatMapFuture(_.msgAndLikes.getMessageAndLikes(id))(Threading.Background).map { m => set(m.getOrElse(MessageAndLikes.Deleted)) } (Threading.Ui)

  def set(msg: MessageAndLikes): Unit = if (msg.message != data) {
    data = msg.message
    notifyChanged()
    convId ! data.convId
    notifyEphemeralRead()
  }


  override def addUpdateListener(listener: UpdateListener): Unit = {
    super.addUpdateListener(listener)
    notifyEphemeralRead()
  }

  private def notifyEphemeralRead() =
    if (data.ephemeral != EphemeralExpiration.NONE && data.expiryTime.isEmpty && getListenersCount > 0)
      context.zms { _.ephemeral.onMessageRead(id) }

  private def content = if (data.content.isEmpty) MessageContent.Empty else data.content.head

  override def retry(): Unit =
    if (data.state == Status.FAILED || data.state == Status.FAILED_READ) context.messages.retry(data.convId, data.id)
    else error(s"Retrying a message that has not yet failed (${data.state}): $id")

  override def equals(other: Any): Boolean = other match {
    case other: Message => id == other.id
    case _ => false
  }

  override def hashCode: Int = id.hashCode

  override def toString: String = s"Message($id, $content, $data)"
}
